// @omneedia/authom/lib/services/letmein.js
var EventEmitter = require("events").EventEmitter
var util = require("util");

function LetMeIn_Auth(options) {

	this.on("request", function (req, res) {
		var _p = this;
		// request
		global.request = function (o, cb) {
			var request = require('request');
			if (global.CFG.current.proxy) var Request = request.defaults({
				'proxy': global.CFG.current.proxy
			});
			else var Request = request;
			return Request(o, cb);
		};
		var OA = {
			authenticate: function (req, res, callback) {
				if (typeof req.query.pid == "undefined") {
					var redirectURL = options.login.uri;
					res.writeHead(307, {
						'Location': redirectURL
					});
					res.write('<a href="' + redirectURL + '">login</a>');
					res.end();
				} else {
					// we collect information
					global.request({
						url: options.login.uri.split('/login')[0] + '/pid',
						method: "post",
						form: {
							pid: req.query.pid
						}
					}, function (err, resp, body) {
						var response = JSON.parse(body.toString('utf-8'));
						callback(null, true, response.username, null);
					});
				}
			}
		};
		OA.authenticate(req, res, function (err, status, username, extended) {
			var profile = {
				service: "letmein",
				username: username
			};
			if (err) {
				_p.emit("error", req, res, err);
			} else {
				_p.emit("auth", req, res, profile)
			}
		});

	});
	EventEmitter.call(this);
};

util.inherits(LetMeIn_Auth, EventEmitter);

module.exports = LetMeIn_Auth;
